designs/disruption-controls-by-action.md

```diff
@@ -0,0 +1,218 @@
+} // ... Same as Existing 
+
+type DisruptionAction string 
+
+const (
+	All DisruptionAction = "All"
+	Consolidation DisruptionAction = "Consolidation" 
+	Drift DisruptionAction = "Drift" 
+	Expiration DisruptionAction = "Expiration" 
+	Emptiness DisruptionAction = "Emptiness" ? 
+)
+```
+
+THis design proposes first defining Action Controls as a very barebones structure with a place in the api  
+```go
+// ActionControls defines the controls for a particular DisruptionAction, these controls are meant to apply for generic controls that apply to all disruption actions. 
+type ActionControls struct {
+	// Budgets is a list of Budgets. 
+	// If there are multiple active budgets, Karpenter uses 
+	// the most restrictive value. If left undefined, 
+	// this will default to one budget with a value to 10%. 
+	// +kubebuilder:validation:XValidation:message="'schedule' must be set with 'duration'",rule="!self.all(x, (has(x.schedule) && !has(x.duration)) || (!has(x.schedule) && has(x.duration)))" 
+	// +kubebuilder:default:={{nodes: "10%"}} 
+	// +kubebuilder:validation:MaxItems=50 
+	// +optional 
+	Budgets []Budget `json:"budgets,omitempty" hash:"ignore"` 
+}
+```
+
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    consolidationPolicy: WhenUnderutilized
+    genericPerActionControls:
+      Consolidation:
+        disruptAfter: "30m"
+        budgets:
+          - nodes: "20%"
+            schedule: "0 0 1 * *"
+            duration: "1h"
+      Drift:
+        disruptAfter: "1h"
+        budgets:
+          - nodes: "10%"
+            schedule: "0 0 * * 0"
+            duration: "2h"
+      Expiration:
+        disruptAfter: "Never"
+```
+
+#### Considerations 
+Some of the API choices for a given action seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each action. It appears there is a need for disruption controls within the budgets for each action, not just overall.
+
+This approach aligns well with controls that apply to all existing actions. The proposal presented here is similar to the one mentioned above in relation to the actions we allow to be defined (All, Consolidation, Drift, Expiration, Emptiness).
+
+This proposal is currently scoped for disruptionBudgets by action. However, we should also consider incorporating other generic disruption controls into the PerActionControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-action controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-action level of granularity.
```

Comment:
This is an interesting point. It might worth be exploring how this has interplay with some other disruption mechanisms that we are talking about like the ability to control whether or not you get a taint added to the node when a given condition is met OR the ability to control a force disruption option that automatically proceeds with disruption after a given time even if there are blocking PDBs or `karpenter.sh/do-not-disrupt` annotations

Reply:
forceDisruption seems interesting and useful for something like upgrade. Do you have a document or something of the like I can take a look at for this?